src/files/normalise-input.js

```diff
@@ -0,0 +1,233 @@
+ * String
+ * Iterable<Number>
+ * Iterable<Buffer>
+ * Iterable<Blob>
+ * Iterable<{ path, content: Buffer }>
+ * Iterable<{ path, content: Blob }>
+ * Iterable<{ path, content: Iterable<Number> }>
+ * Iterable<{ path, content: AsyncIterable<Buffer> }>
+ * AsyncIterable<Buffer>
+ * AsyncIterable<{ path, content: Buffer }>
+ * AsyncIterable<{ path, content: Blob }>
+ * AsyncIterable<{ path, content: Iterable<Buffer> }>
+ * AsyncIterable<{ path, content: AsyncIterable<Buffer> }>
+ * ```
+ * Into:
+ *
+ * ```
+ * AsyncIterable<{ path, content: AsyncIterable<Buffer> }>
+ * ```
+ *
+ * @param input Object
+ * @return AsyncInterable<{ path, content: AsyncIterable<Buffer> }>
+ */
+module.exports = function normaliseInput (input) {
+  // must give us something
+  if (input === null || input === undefined) {
+    throw errCode(new Error(`Unexpected input: ${input}`, 'ERR_UNEXPECTED_INPUT'))
+  }
+
+  // { path, content: ? }
+  if (isFileObject(input)) {
+    return (async function * () { // eslint-disable-line require-await
+      yield toFileObject(input)
+    })()
+  }
+
+  // String
+  if (typeof input === 'string' || input instanceof String) {
+    return (async function * () { // eslint-disable-line require-await
+      yield toFileObject(input)
+    })()
+  }
+
+  // Buffer|ArrayBuffer|TypedArray
+  // Blob|File
+  if (isBytes(input) || isBloby(input)) {
+    return (async function * () { // eslint-disable-line require-await
+      yield toFileObject(input)
+    })()
+  }
+
+  // Iterable<?>
+  if (input[Symbol.iterator]) {
+    // Iterable<Number>
+    if (!isNaN(input[0])) {
+      return (async function * () { // eslint-disable-line require-await
+        yield toFileObject([input])
+      })()
+    }
+
+    // Iterable<Buffer>
+    // Iterable<Blob>
+    return (async function * () { // eslint-disable-line require-await
+      for (const chunk of input) {
+        yield toFileObject(chunk)
+      }
+    })()
+  }
+
+  // AsyncIterable<?>
+  if (input[Symbol.asyncIterator]) {
+    return (async function * () { // eslint-disable-line require-await
+      for await (const chunk of input) {
+        yield toFileObject(chunk)
+      }
+    })()
+  }
+
+  throw errCode(new Error('Unexpected input: ' + typeof input), 'ERR_UNEXPECTED_INPUT')
+}
+
+function toFileObject (input) {
+  return {
+    path: input.path || '',
+    content: toAsyncIterable(input.content || input)
+  }
+}
+
+function toAsyncIterable (input) {
+  // Buffer|ArrayBuffer|TypedArray|array of bytes
+  if (isBytes(input)) {
+    return (async function * () { // eslint-disable-line require-await
+      yield toBuffer(input)
+    })()
+  }
+
+  if (typeof input === 'string' || input instanceof String) {
+    return (async function * () { // eslint-disable-line require-await
+      yield toBuffer(input)
+    })()
+  }
+
+  // Blob|File
+  if (isBloby(input)) {
+    return blobToAsyncGenerator(input)
+  }
+
+  // Iterator<?>
+  if (input[Symbol.iterator]) {
+    if (!isNaN(input[0])) {
```

Comment:
Not is not a number means it's a number.  Of course it'll happily accept floats and negative numbers, etc which would lead to 🤪.  `Number.isInteger` is probably better in this case..

I'd hoped to avoid calling `.next()` on the iterator in case it consumes a value that otherwise wouldn't be re-iterated - e.g. more of a `.peek()`-type mechanism.  Could probably wrap the iterator in another iterator(!) that does something a bit more sensible.

Reply:
`it-peek`!

Reply:
> Not is not a number means it's a number 😉 . Of course it'll happily accept floats and negative numbers, etc which would lead to 🤪.

😉 ...and other values are are not `NaN`. `isNaN` is just the worst. For the benefit of anyone else reading this the following non-numbers would also make it through!:

```js
!isNaN(null) // true
!isNaN('123') // true
!isNaN(' ') // true
!isNaN([]) // true
!isNaN(new Date) // true
```